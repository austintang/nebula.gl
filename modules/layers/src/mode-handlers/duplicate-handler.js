// @flow

import type { StartDraggingEvent } from '../event-types.js';
import type { EditAction } from './mode-handler.js';
import { SnapTranslateHandler } from './snap-translate-handler';

export class DuplicateHandler extends SnapTranslateHandler {
  handleStartDragging(event: StartDraggingEvent): ?EditAction {
    if (!this._isTranslatable) {
      return null;
    }

    this._geometryBeforeTranslate = this.getSelectedFeaturesAsFeatureCollection();

    return this._geometryBeforeTranslate
      ? this.getAddManyFeaturesAction(this._geometryBeforeTranslate)
      : null;
  }

  getCursor({ isDragging }: { isDragging: boolean }): string {
    if (this._isTranslatable) {
      return 'copy';
    }
    return isDragging ? 'grabbing' : 'grab';
  }
}
