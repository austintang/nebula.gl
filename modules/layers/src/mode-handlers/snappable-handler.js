// @flow

import type { FeatureCollection, Position } from '../geojson-types.js';
import type { PointerMoveEvent, StartDraggingEvent, StopDraggingEvent } from '../event-types.js';
import type { EditHandle, EditAction } from './mode-handler.js';
import { ModeHandler, getPickedEditHandle, getEditHandlesForGeometry } from './mode-handler';

const DEFAULT_SNAP_PIXELS = 5;

type HandlePicks = { pickedHandle?: EditHandle, potentialSnapHandle?: EditHandle };

export class SnappableHandler extends ModeHandler {
  _handler: ModeHandler;
  _editHandlePicks: ?HandlePicks;
  _isSnapped: boolean;

  constructor(handler: ModeHandler) {
    super();
    this._handler = handler;
  }

  setFeatureCollection(featureCollection: FeatureCollection): void {
    this._handler.setFeatureCollection(featureCollection);
  }

  setModeConfig(modeConfig: any): void {
    this._handler.setModeConfig(modeConfig);
  }

  setSelectedFeatureIndexes(indexes: number[]): void {
    this._handler.setSelectedFeatureIndexes(indexes);
  }

  setDeckGlContext(context: Object) {
    this._handler.setDeckGlContext(context);
  }

  _getSnappedMouseEvent(event: Object, snapPoint: Position): PointerMoveEvent {
    const snapPointScreenCoords = this._handler._context.viewport.project(snapPoint);
    event.groundCoords = snapPoint;
    event.screenCoords = snapPointScreenCoords;

    return event;
  }

  _getEditHandlePicks(event: PointerMoveEvent): HandlePicks {
    const { screenCoords } = event;
    const { snapPixels } = this._handler.getModeConfig() || {};

    const picks = this._handler._context.layerManager.pickObject({
      x: screenCoords[0],
      y: screenCoords[1],
      mode: 'query',
      layerIds: ['geojson-point-edit-handles'],
      radius: snapPixels || DEFAULT_SNAP_PIXELS,
      viewports: [this._handler._context.viewport],
      depth: 2
    });

    const potentialSnapHandle = picks.find(
      pick => pick.object && pick.object.type === 'intermediate'
    );
    const handles = { potentialSnapHandle: potentialSnapHandle && potentialSnapHandle.object };

    const pickedHandle = getPickedEditHandle(event.pointerDownPicks);
    if (pickedHandle) {
      return { ...handles, pickedHandle };
    }

    return handles;
  }

  _updatePickedHandlePosition(editAction: EditAction) {
    const { pickedHandle } = this._editHandlePicks || {};

    if (pickedHandle && editAction) {
      const { featureIndexes, updatedData } = editAction;

      for (let i = 0; i < featureIndexes.length; i++) {
        const selectedIndex = featureIndexes[i];
        const updatedFeature = updatedData.features[selectedIndex];

        const { positionIndexes, featureIndex } = pickedHandle;
        if (selectedIndex >= 0 && featureIndex !== selectedIndex) return;

        const { coordinates } = updatedFeature.geometry;
        // $FlowFixMe
        (pickedHandle || {}).position = positionIndexes.reduce(
          (a: any[], b: number) => a[b],
          coordinates
        );
      }
    }
  }

  _getNonPickedIntermediateHandles(): EditHandle[] {
    const handles = [];
    const { features } = this._handler.featureCollection.getObject();

    for (let i = 0; i < features.length; i++) {
      if (!this._handler.getSelectedFeatureIndexes().includes(i) && i < features.length) {
        const { geometry } = features[i];
        handles.push(...getEditHandlesForGeometry(geometry, i, 'intermediate'));
      }
    }
    return handles;
  }

  // If no snap handle has been picked, only display the edit handles of the
  // selected feature. If a snap handle has been picked, display said snap handle
  // along with all snappable points on all non-selected features.
  getEditHandles(picks?: Array<Object>, groundCoords?: Position): any[] {
    const { enableSnapping } = this._handler.getModeConfig() || {};
    const handles = this._handler.getEditHandles(picks, groundCoords);

    if (!enableSnapping) return handles;
    const { pickedHandle } = this._editHandlePicks || {};

    if (pickedHandle) {
      handles.push(...this._getNonPickedIntermediateHandles(), pickedHandle);
      return handles;
    }

    const { features } = this._handler.featureCollection.getObject();
    for (const index of this._handler.getSelectedFeatureIndexes()) {
      if (index < features.length) {
        const { geometry } = features[index];
        handles.push(...getEditHandlesForGeometry(geometry, index, 'snap'));
      }
    }

    return handles.filter(Boolean);
  }

  _shouldPerformSnap() {
    if (!this._editHandlePicks) return false;
    const { pickedHandle, potentialSnapHandle } = this._editHandlePicks || {};
    return pickedHandle && potentialSnapHandle;
  }

  _performSnapIfRequired() {
    if (this._isSnapped || !this._editHandlePicks || !this._shouldPerformSnap()) return;
    const { pickedHandle, potentialSnapHandle } = this._editHandlePicks || {};
    if (!pickedHandle || !potentialSnapHandle) return;

    this._isSnapped = true;
  }

  // Unsnapping only occurs after the user snaps two polygons but continues to drag the
  // cursor past the point of resistance.
  _performUnsnapIfRequired() {
    if (!this._isSnapped) return;

    const { potentialSnapHandle } = this._editHandlePicks || {};
    if (!potentialSnapHandle) {
      this._isSnapped = false;
    }
  }

  _getSnapAwareEvent(event: Object): Object {
    const { potentialSnapHandle } = this._editHandlePicks || {};

    return potentialSnapHandle && potentialSnapHandle.position
      ? this._getSnappedMouseEvent(event, potentialSnapHandle.position)
      : event;
  }

  handleStartDragging(event: StartDraggingEvent): ?EditAction {
    return this._handler.handleStartDragging(event);
  }

  handleStopDragging(event: StopDraggingEvent): ?EditAction {
    const modeActionSummary = this._handler.handleStopDragging(this._getSnapAwareEvent(event));

    this._editHandlePicks = null;
    this._isSnapped = false;
    return modeActionSummary;
  }

  getCursor(event: { isDragging: boolean }): string {
    return this._handler.getCursor(event);
  }

  handlePointerMove(event: PointerMoveEvent): { editAction: ?EditAction, cancelMapPan: boolean } {
    this._editHandlePicks = this._getEditHandlePicks(event);
    const { enableSnapping } = this._handler.getModeConfig() || {};

    if (enableSnapping && this._editHandlePicks) {
      this._performSnapIfRequired();
      this._performUnsnapIfRequired();
    }

    const modeActionSummary = this._handler.handlePointerMove(this._getSnapAwareEvent(event));
    const { editAction } = modeActionSummary;
    if (editAction) {
      this._updatePickedHandlePosition(editAction);
    }

    return modeActionSummary;
  }
}
