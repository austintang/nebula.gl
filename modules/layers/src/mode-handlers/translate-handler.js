// @flow

import turfBearing from '@turf/bearing';
import turfDistance from '@turf/distance';
import { point as turfPoint } from '@turf/helpers';
import turfTransformTranslate from '@turf/transform-translate';
import type { FeatureCollection, Position } from '../geojson-types.js';
import type { PointerMoveEvent, StartDraggingEvent, StopDraggingEvent } from '../event-types.js';
import { ImmutableFeatureCollection } from '../immutable-feature-collection.js';
import type { EditAction } from './mode-handler.js';
import { getPickedEditHandle } from './mode-handler';
import { SnapHandler } from './snap-handler.js';

export class TranslateHandler extends SnapHandler {
  _geometryBeforeTranslate: ?FeatureCollection;
  _isTranslatable: boolean;
  _unsnapMousePointStart: Position;
  _updatedData: ImmutableFeatureCollection;

  handlePointerMove(event: PointerMoveEvent): { editAction: ?EditAction, cancelMapPan: boolean } {
    this.setPickedHandle(getPickedEditHandle(event.pointerDownPicks));

    let editAction: ?EditAction = null;

    this._isTranslatable =
      Boolean(this._geometryBeforeTranslate) || this.isSelectionPicked(event.picks);

    if (!this._isTranslatable || !event.pointerDownGroundCoords) {
      // Nothing to do
      return { editAction: null, cancelMapPan: false };
    }

    if (event.isDragging && this._geometryBeforeTranslate) {
      // Translate the geometry
      editAction = this.getTranslateAction(
        event.pointerDownGroundCoords,
        event.groundCoords,
        'translating'
      );
    }

    return { editAction, cancelMapPan: true };
  }

  handleStartDragging(event: StartDraggingEvent): ?EditAction {
    if (!this._isTranslatable) {
      return null;
    }

    this._geometryBeforeTranslate = this.getSelectedFeaturesAsFeatureCollection();
    this._unsnapMousePointStart = event.pointerDownGroundCoords;
    return null;
  }

  handleStopDragging(event: StopDraggingEvent): ?EditAction {
    this.clearCachedHandles();
    let editAction: ?EditAction = null;

    if (this._geometryBeforeTranslate) {
      // Translate the geometry
      editAction = this.getTranslateAction(
        event.pointerDownGroundCoords,
        event.groundCoords,
        'translated'
      );
      this._geometryBeforeTranslate = null;
    }

    return editAction;
  }

  getCursor({ isDragging }: { isDragging: boolean }): string {
    if (this._isTranslatable) {
      return 'move';
    }
    return isDragging ? 'grabbing' : 'grab';
  }

  calculateDistanceAndDirection(startDragPoint: Position, currentPoint: Position) {
    const p1 = turfPoint(startDragPoint);
    const p2 = turfPoint(currentPoint);

    const distanceMoved = turfDistance(p1, p2);
    const direction = turfBearing(p1, p2);

    return {
      distanceMoved,
      direction
    };
  }

  performSnap(currentPoint: Position, snapStrength: number) {
    if (this._potentialNonPickedHandle) {
      const snapDetails = this.calculateSnapIfWithinThreshold(snapStrength);
      if (snapDetails && snapDetails.length) {
        for (const snapDetail of snapDetails) {
          const { movedFeature, selectedIndex } = snapDetail;

          this._updatedData = this._updatedData.replaceGeometry(
            selectedIndex,
            movedFeature.geometry
          );

          this.updatePickedHandlePosition(selectedIndex, movedFeature.geometry);
          this._unsnapMousePointStart = currentPoint;
        }
      }
    }
  }

  performUnsnap(currentPoint: Position, snapStrength: number) {
    const { distanceMoved: unsnapDistanceMoved } = this.calculateDistanceAndDirection(
      this._unsnapMousePointStart,
      currentPoint
    );

    const unsnapStrengthModifier = this.getSnapStrengthModifier();
    const unsnapStrength = snapStrength * 2;
    if (unsnapDistanceMoved >= unsnapStrength * unsnapStrengthModifier) {
      const selectedIndexes = this.getSelectedFeatureIndexes();

      selectedIndexes.forEach(index =>
        this.getSnapAssociates(index).forEach(associateIndex => {
          if (!selectedIndexes.includes(associateIndex)) {
            this.clearSnapAssociates(index, associateIndex);
          }
        })
      );
    }
  }

  performTranslate(distanceMoved: number, direction: number) {
    const movedFeatures = turfTransformTranslate(
      this._geometryBeforeTranslate,
      distanceMoved,
      direction
    );

    const selectedIndexes = this.getSelectedFeatureIndexes();
    for (let i = 0; i < selectedIndexes.length; i++) {
      const selectedIndex = selectedIndexes[i];
      const movedFeature = movedFeatures.features[i];
      this._updatedData = this._updatedData.replaceGeometry(selectedIndex, movedFeature.geometry);
      this.updatePickedHandlePosition(selectedIndex, movedFeature.geometry);
    }
  }

  getTranslateAction(
    startDragPoint: Position,
    currentPoint: Position,
    editType: string
  ): ?EditAction {
    if (!this._geometryBeforeTranslate) {
      return null;
    }

    const { snapStrength } = this.getModeConfig() || {};
    const selectedIndexes = this.getSelectedFeatureIndexes();

    const { distanceMoved, direction } = this.calculateDistanceAndDirection(
      startDragPoint,
      currentPoint
    );
    this._updatedData = this.getImmutableFeatureCollection();

    // Perform snap
    if (this.shouldPerformSnap()) {
      this.performSnap(currentPoint, snapStrength);
    }

    // Perform unsnap
    if (this.shouldPerformUnsnap()) {
      this.performUnsnap(currentPoint, snapStrength);
    }

    // Perform standard translate
    if (this.shouldPerformStandardModeAction()) {
      this.performTranslate(distanceMoved, direction);
    }

    return {
      updatedData: this._updatedData.getObject(),
      editType,
      featureIndexes: selectedIndexes,
      editContext: null
    };
  }
}
