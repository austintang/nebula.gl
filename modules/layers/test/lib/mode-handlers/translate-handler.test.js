// @flow
import { TranslateHandler } from '../../../src/mode-handlers/translate-handler';
import type { FeatureCollection } from '../../../src/geojson-types.js';
import {
  createFeatureCollection,
  featuresForSnappingTests,
  mockPickedHandle,
  mockNonPickedHandle
} from '../test-utils.js';

import {
  testModeHandlerHandlePointMove,
  testHandleStartDragging,
  testHandleStopDragging
} from './mode-handler-utils';

const modeName = 'translate';
const featureCollection: FeatureCollection = createFeatureCollection({
  mockGeoJsonProperties: true
});

function mockFeatureCollectionState(features: any) {
  const handler = new TranslateHandler(features);
  handler.setSelectedFeatureIndexes([1]);
  return handler;
}

testModeHandlerHandlePointMove(modeName, TranslateHandler, featureCollection, '_isTranslatable');

testHandleStartDragging(
  modeName,
  TranslateHandler,
  featureCollection,
  '_isTranslatable',
  '_geometryBeforeTranslate'
);

testHandleStopDragging(modeName, TranslateHandler, featureCollection);

describe('translate-handler specific functions', () => {
  let handler;

  beforeEach(() => {
    handler = mockFeatureCollectionState(featuresForSnappingTests);
  });

  test('getCursor() - _isTranslatable is true', () => {
    handler._isTranslatable = true;
    const getCursorReturn = handler.getCursor({ isDragging: false });
    expect(getCursorReturn).toEqual('move');
  });

  test('getCursor() - _isTranslatable is false and isDragging is true', () => {
    handler._isTranslatable = false;
    const getCursorReturn = handler.getCursor({ isDragging: true });
    expect(getCursorReturn).toEqual('grabbing');
  });

  test('getCursor() - _isTranslatable is false and isDragging is false', () => {
    handler._isTranslatable = false;
    const getCursorReturn = handler.getCursor({ isDragging: false });
    expect(getCursorReturn).toEqual('grab');
  });

  test('calculateDistanceAndDirection()', () => {
    const distanceAndDirection = handler.calculateDistanceAndDirection([1, 1], [100, 100]);
    expect(distanceAndDirection).toMatchSnapshot();
  });

  test('performSnap()', () => {
    handler.setModeConfig({ enablePolygonSnapping: true });
    handler.setSelectedFeatureIndexes([0]);
    handler._pickedHandle = mockPickedHandle;
    handler._potentialNonPickedHandle = mockNonPickedHandle;
    handler._updatedData = handler.getImmutableFeatureCollection();

    handler.shouldPerformSnap();
    const mockCurrentPoint = [-122.49485401280788, 37.987923255302974];
    handler.performSnap(mockCurrentPoint, 10);

    expect(handler._snapAssociations).toEqual([[1], [0]]);
    expect(handler._unsnapMousePointStart).toEqual(mockCurrentPoint);
    expect(handler).toMatchSnapshot();
  });

  test('performUnsnap()', () => {
    handler.setModeConfig({ enablePolygonSnapping: true });
    handler.setSelectedFeatureIndexes([0]);
    handler._pickedHandle = mockPickedHandle;
    handler._potentialNonPickedHandle = mockNonPickedHandle;
    handler._updatedData = handler.getImmutableFeatureCollection();
    const mockCurrentPoint = [-122.49485401280788, 37.987923255302974];
    handler._unsnapMousePointStart = [1, 1];
    handler._snapAssociations = [[1], [0]];

    handler.performUnsnap(mockCurrentPoint, 1);
    expect(handler._snapAssociations).toEqual([[], []]);
    expect(handler).toMatchSnapshot();
  });

  test('performTranslate()', () => {
    const selectedIndex = 0;
    handler.setSelectedFeatureIndexes([selectedIndex]);
    handler._updatedData = handler.getImmutableFeatureCollection();
    handler._geometryBeforeTranslate = handler.getSelectedFeaturesAsFeatureCollection();

    const initialFeatures = handler._updatedData.getObject().features;
    const initialNonSelectedFeatures = initialFeatures.filter(
      (_, index) => index !== selectedIndex
    );

    handler.performTranslate(100, 20);
    const featuresAfterTranslate = handler._updatedData.getObject().features;
    const nonSelectedFeaturesAfterTranslate = featuresAfterTranslate.filter(
      (_, index) => index !== selectedIndex
    );

    expect(featuresAfterTranslate).not.toEqual(initialFeatures);
    expect(nonSelectedFeaturesAfterTranslate).toEqual(initialNonSelectedFeatures);
  });

  test('getTranslateAction()', () => {
    handler.setSelectedFeatureIndexes([0]);
    handler._updatedData = handler.getImmutableFeatureCollection();
    handler._geometryBeforeTranslate = handler.getSelectedFeaturesAsFeatureCollection();

    const translateAction = handler.getTranslateAction([1, 1], [100, 100], 'test-edit');
    expect(translateAction).toMatchSnapshot();
  });
});
