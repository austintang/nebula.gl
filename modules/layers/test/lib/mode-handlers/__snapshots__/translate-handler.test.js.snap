// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translate-handler specific functions calculateDistanceAndDirection() 1`] = `
Object {
  "direction": -9.885464830954206,
  "distanceMoved": 9724.92501810167,
}
`;

exports[`translate-handler specific functions getTranslateAction() 1`] = `
Object {
  "editContext": null,
  "editType": "test-edit",
  "featureIndexes": Array [
    0,
  ],
  "updatedData": Object {
    "features": Array [
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              Array [
                -127.11509550096605,
                55.852345878852674,
              ],
              Array [
                -127.06276731340756,
                55.852345878852674,
              ],
              Array [
                -127.08892433247462,
                55.90153707629159,
              ],
              Array [
                -127.14125252003316,
                55.90153707629159,
              ],
              Array [
                -127.11509550096605,
                55.852345878852674,
              ],
            ],
          ],
          "type": "Polygon",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              Array [
                -122.40814940182412,
                37.97528325161274,
              ],
              Array [
                -122.37737022011595,
                37.97528325161274,
              ],
              Array [
                -122.37737142575622,
                37.952934704562644,
              ],
              Array [
                -122.40815060746439,
                37.952934704562644,
              ],
              Array [
                -122.40814940182412,
                37.97528325161274,
              ],
            ],
          ],
          "type": "Polygon",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            -122.28103267622373,
            37.98843664327903,
          ],
          "type": "Point",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              Array [
                -122.40908525092362,
                37.85902221692065,
              ],
              Array [
                -122.34432261530361,
                37.85902221692065,
              ],
              Array [
                -122.3442469760231,
                37.8157115979288,
              ],
              Array [
                -122.40900961164311,
                37.8157115979288,
              ],
              Array [
                -122.40908525092362,
                37.85902221692065,
              ],
            ],
          ],
          "type": "Polygon",
        },
        "properties": Object {},
        "type": "Feature",
      },
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              Array [
                -122.40682264287454,
                38.13330760982133,
              ],
              Array [
                -122.34143228624993,
                38.13330760982133,
              ],
              Array [
                -122.34144710122638,
                38.08906337516829,
              ],
              Array [
                -122.40683745785105,
                38.08906337516829,
              ],
              Array [
                -122.40682264287454,
                38.13330760982133,
              ],
            ],
          ],
          "type": "Polygon",
        },
        "properties": Object {},
        "type": "Feature",
      },
    ],
    "type": "FeatureCollection",
  },
}
`;

exports[`translate-handler specific functions performSnap() 1`] = `
TranslateHandler {
  "_clickSequence": Array [],
  "_context": undefined,
  "_geometryBeforeTranslate": undefined,
  "_isTranslatable": undefined,
  "_modeConfig": Object {
    "enablePolygonSnapping": true,
  },
  "_pickedHandle": Object {
    "featureIndex": 0,
    "position": Array [
      -122.40814192776702,
      37.97531507849716,
    ],
    "positionIndexes": Array [
      0,
      0,
    ],
    "type": "snap",
  },
  "_potentialNonPickedHandle": Object {
    "featureIndex": 1,
    "position": Array [
      -122.40814940182412,
      37.97528325161274,
    ],
    "positionIndexes": Array [
      0,
      0,
    ],
    "type": "intermediate",
  },
  "_potentialSnapDistance": 7.728100362611004,
  "_selectedFeatureIndexes": Array [
    0,
  ],
  "_snapAssociations": Array [
    Array [
      1,
    ],
    Array [
      0,
    ],
  ],
  "_tentativeFeature": null,
  "_unsnapMousePointStart": Array [
    -122.49485401280788,
    37.987923255302974,
  ],
  "_updatedData": ImmutableFeatureCollection {
    "featureCollection": Object {
      "features": Array [
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.40814192776702,
                  37.97531507849716,
                ],
                Array [
                  -122.35581374020853,
                  37.97531507849716,
                ],
                Array [
                  -122.35587444851501,
                  37.92612388105825,
                ],
                Array [
                  -122.40820263607355,
                  37.92612388105825,
                ],
                Array [
                  -122.40814192776702,
                  37.97531507849716,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.40814940182412,
                  37.97528325161274,
                ],
                Array [
                  -122.37737022011595,
                  37.97528325161274,
                ],
                Array [
                  -122.37737142575622,
                  37.952934704562644,
                ],
                Array [
                  -122.40815060746439,
                  37.952934704562644,
                ],
                Array [
                  -122.40814940182412,
                  37.97528325161274,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              -122.28103267622373,
              37.98843664327903,
            ],
            "type": "Point",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.40908525092362,
                  37.85902221692065,
                ],
                Array [
                  -122.34432261530361,
                  37.85902221692065,
                ],
                Array [
                  -122.3442469760231,
                  37.8157115979288,
                ],
                Array [
                  -122.40900961164311,
                  37.8157115979288,
                ],
                Array [
                  -122.40908525092362,
                  37.85902221692065,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.40682264287454,
                  38.13330760982133,
                ],
                Array [
                  -122.34143228624993,
                  38.13330760982133,
                ],
                Array [
                  -122.34144710122638,
                  38.08906337516829,
                ],
                Array [
                  -122.40683745785105,
                  38.08906337516829,
                ],
                Array [
                  -122.40682264287454,
                  38.13330760982133,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
      ],
      "type": "FeatureCollection",
    },
  },
  "featureCollection": ImmutableFeatureCollection {
    "featureCollection": Object {
      "features": Array [
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.49485401280788,
                  37.987923255302974,
                ],
                Array [
                  -122.44252582524939,
                  37.987923255302974,
                ],
                Array [
                  -122.44252847887157,
                  37.93873205786406,
                ],
                Array [
                  -122.49485666643005,
                  37.93873205786406,
                ],
                Array [
                  -122.49485401280788,
                  37.987923255302974,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.40814940182412,
                  37.97528325161274,
                ],
                Array [
                  -122.37737022011595,
                  37.97528325161274,
                ],
                Array [
                  -122.37737142575622,
                  37.952934704562644,
                ],
                Array [
                  -122.40815060746439,
                  37.952934704562644,
                ],
                Array [
                  -122.40814940182412,
                  37.97528325161274,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              -122.28103267622373,
              37.98843664327903,
            ],
            "type": "Point",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.40908525092362,
                  37.85902221692065,
                ],
                Array [
                  -122.34432261530361,
                  37.85902221692065,
                ],
                Array [
                  -122.3442469760231,
                  37.8157115979288,
                ],
                Array [
                  -122.40900961164311,
                  37.8157115979288,
                ],
                Array [
                  -122.40908525092362,
                  37.85902221692065,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.40682264287454,
                  38.13330760982133,
                ],
                Array [
                  -122.34143228624993,
                  38.13330760982133,
                ],
                Array [
                  -122.34144710122638,
                  38.08906337516829,
                ],
                Array [
                  -122.40683745785105,
                  38.08906337516829,
                ],
                Array [
                  -122.40682264287454,
                  38.13330760982133,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
      ],
      "type": "FeatureCollection",
    },
  },
}
`;

exports[`translate-handler specific functions performUnsnap() 1`] = `
TranslateHandler {
  "_clickSequence": Array [],
  "_context": undefined,
  "_geometryBeforeTranslate": undefined,
  "_isTranslatable": undefined,
  "_modeConfig": Object {
    "enablePolygonSnapping": true,
  },
  "_pickedHandle": Object {
    "featureIndex": 0,
    "position": Array [
      -122.40814192776702,
      37.97531507849716,
    ],
    "positionIndexes": Array [
      0,
      0,
    ],
    "type": "snap",
  },
  "_potentialNonPickedHandle": Object {
    "featureIndex": 1,
    "position": Array [
      -122.37737022011595,
      37.97528325161274,
    ],
    "positionIndexes": Array [
      0,
      1,
    ],
    "type": "intermediate",
  },
  "_potentialSnapDistance": undefined,
  "_selectedFeatureIndexes": Array [
    0,
  ],
  "_snapAssociations": Array [
    Array [],
    Array [],
  ],
  "_tentativeFeature": null,
  "_unsnapMousePointStart": Array [
    1,
    1,
  ],
  "_updatedData": ImmutableFeatureCollection {
    "featureCollection": Object {
      "features": Array [
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.49485401280788,
                  37.987923255302974,
                ],
                Array [
                  -122.44252582524939,
                  37.987923255302974,
                ],
                Array [
                  -122.44252847887157,
                  37.93873205786406,
                ],
                Array [
                  -122.49485666643005,
                  37.93873205786406,
                ],
                Array [
                  -122.49485401280788,
                  37.987923255302974,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.40814940182412,
                  37.97528325161274,
                ],
                Array [
                  -122.37737022011595,
                  37.97528325161274,
                ],
                Array [
                  -122.37737142575622,
                  37.952934704562644,
                ],
                Array [
                  -122.40815060746439,
                  37.952934704562644,
                ],
                Array [
                  -122.40814940182412,
                  37.97528325161274,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              -122.28103267622373,
              37.98843664327903,
            ],
            "type": "Point",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.40908525092362,
                  37.85902221692065,
                ],
                Array [
                  -122.34432261530361,
                  37.85902221692065,
                ],
                Array [
                  -122.3442469760231,
                  37.8157115979288,
                ],
                Array [
                  -122.40900961164311,
                  37.8157115979288,
                ],
                Array [
                  -122.40908525092362,
                  37.85902221692065,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.40682264287454,
                  38.13330760982133,
                ],
                Array [
                  -122.34143228624993,
                  38.13330760982133,
                ],
                Array [
                  -122.34144710122638,
                  38.08906337516829,
                ],
                Array [
                  -122.40683745785105,
                  38.08906337516829,
                ],
                Array [
                  -122.40682264287454,
                  38.13330760982133,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
      ],
      "type": "FeatureCollection",
    },
  },
  "featureCollection": ImmutableFeatureCollection {
    "featureCollection": Object {
      "features": Array [
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.49485401280788,
                  37.987923255302974,
                ],
                Array [
                  -122.44252582524939,
                  37.987923255302974,
                ],
                Array [
                  -122.44252847887157,
                  37.93873205786406,
                ],
                Array [
                  -122.49485666643005,
                  37.93873205786406,
                ],
                Array [
                  -122.49485401280788,
                  37.987923255302974,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.40814940182412,
                  37.97528325161274,
                ],
                Array [
                  -122.37737022011595,
                  37.97528325161274,
                ],
                Array [
                  -122.37737142575622,
                  37.952934704562644,
                ],
                Array [
                  -122.40815060746439,
                  37.952934704562644,
                ],
                Array [
                  -122.40814940182412,
                  37.97528325161274,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              -122.28103267622373,
              37.98843664327903,
            ],
            "type": "Point",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.40908525092362,
                  37.85902221692065,
                ],
                Array [
                  -122.34432261530361,
                  37.85902221692065,
                ],
                Array [
                  -122.3442469760231,
                  37.8157115979288,
                ],
                Array [
                  -122.40900961164311,
                  37.8157115979288,
                ],
                Array [
                  -122.40908525092362,
                  37.85902221692065,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                Array [
                  -122.40682264287454,
                  38.13330760982133,
                ],
                Array [
                  -122.34143228624993,
                  38.13330760982133,
                ],
                Array [
                  -122.34144710122638,
                  38.08906337516829,
                ],
                Array [
                  -122.40683745785105,
                  38.08906337516829,
                ],
                Array [
                  -122.40682264287454,
                  38.13330760982133,
                ],
              ],
            ],
            "type": "Polygon",
          },
          "properties": Object {},
          "type": "Feature",
        },
      ],
      "type": "FeatureCollection",
    },
  },
}
`;
